pkg/fleets/controller.go

```diff
@@ -485,10 +536,27 @@ func (c *Controller) rollingUpdateRest(fleet *agonesv1.Fleet, rest []*agonesv1.G
 			break
 		}
 		gsSetCopy := gsSet.DeepCopy()
+		newReplicasCount := fleet.LowerBoundReplicas(gsSetCopy.Spec.Replicas - unavailable)
 		if gsSet.Status.ShutdownReplicas == 0 {
-			gsSetCopy.Spec.Replicas = fleet.LowerBoundReplicas(gsSetCopy.Spec.Replicas - unavailable)
+			var scaleDownCount int32
+			if runtime.FeatureEnabled(runtime.FeatureRollingUpdateOnReady) {
+				// Wait for new GameServers to become Ready before scaling down Inactive GameServerset
+				// Scale down.
+				scaleDownCount := int32(integer.IntMin(int(gsSet.Spec.Replicas), int(totalScaleDownCount-totalScaledDown)))
+
+				newReplicasCount = gsSet.Spec.Replicas - scaleDownCount
+				if newReplicasCount > gsSet.Spec.Replicas {
+					return fmt.Errorf("when scaling down old GameServerSet, got invalid request to scale down %s/%s %d -> %d", gsSet.Namespace, gsSet.Name, gsSet.Spec.Replicas, newReplicasCount)
```

Comment:
Could you please name those parameters in the log message? It will be much easier to understand these logs.